cmd/pileup/pileup.go

```diff
@@ -0,0 +1,116 @@
+// Command Group: "SAM Tools"
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/sam"
+	"log"
+	"strings"
+)
+
+func usage() {
+	fmt.Print(
+		"pileup - Count bases from sequencing data\n\n" +
+			"pileup prints to stdout by default, but can be redirected with the -o flag.\n" +
+			"The output format is a tab-separated file with columns defined as:\n" +
+			"Chr\tPos\t#A\t#C\t#G\t#T\t#N\t#DEL\tINS_array\n" +
+			"INS_array is an array of observed insertions with format insSeq:readCount\n\n" +
+			"Usage:\n" +
+			"  pileup [options] in.bam\n\n" +
+			"Options:\n")
+	flag.PrintDefaults()
+}
+
+type Settings struct {
+	minDp int
+}
+
+// pileFilters parses the Settings struct to make functions to filter the output piles.
+func pileFilters(s Settings) []func(pile sam.Pile) bool {
+	var filters []func(pile sam.Pile) bool
+
+	if s.minDp > 0 {
+		filters = append(filters, func(pile sam.Pile) bool {
+			var count int
+			for i := range pile.Count {
+				count += pile.Count[i]
+			}
+			for _, i := range pile.InsCount {
+				count += i
+			}
+			return count >= s.minDp
+		})
+	}
+
+	return filters
+}
+
+func pileup(infile string, outfile string, settings Settings) {
+	sBuilder := new(strings.Builder)
+	reads, recycle, header := sam.GoReadToChanRecycle(infile, 1000)
+	sendChan := make(chan sam.Sam) // no buffer to satisfy recycle contract
+
+	// goroutine sends from the read channel to an unbuffered channel
+	// then returns the previous struct to the recycle channel.
+	go func(<-chan *sam.Sam, chan<- *sam.Sam, chan<- sam.Sam) {
```

Comment:
It would be easier to have the pileup function take the read and recycle channels, but I did not want to lock people into using the recycle version of GoReadToChan. This was kind of a middle ground where we got a lot of the benefit of recycling, but could still use the normal GoReadToChan which is a `chan Sam` instead of a `chan *Sam`.

One thing on my to-do list is to rework the GoReadToChanRecycle functions. I wrote them to pass pointers to structs because I wanted to reuse as many fields as possible, but because the largest fields (and the ones most likely to cause GC issues) are already passes as references, it may make mores sense to have the recycle functions operate with chans of structs rather than pointers to get around the issue in this PR of designing for both cases.